crates/ruff/src/rules/airflow/mod.rs

```diff
@@ -0,0 +1,69 @@
+//! Airflow-specific rules
+pub(crate) mod rules;
+
+#[cfg(test)]
+mod tests {
+    use std::path::Path;
+
+    use rustpython_parser::lexer::LexResult;
+    use test_case::test_case;
+    use textwrap::dedent;
+
+    use ruff_python_ast::source_code::{Indexer, Locator, Stylist};
+
+    use crate::linter::{check_path, LinterResult};
+    use crate::registry::{AsRule, Linter, Rule};
+    use crate::settings::flags;
+    use crate::{directives, settings};
+
+    fn rule_code(contents: &str, expected: &[Rule]) {
+        let contents = dedent(contents);
+        let settings = settings::Settings::for_rules(&Linter::Airflow);
+        let tokens: Vec<LexResult> = ruff_rustpython::tokenize(&contents);
+        let locator = Locator::new(&contents);
+        let stylist = Stylist::from_tokens(&tokens, &locator);
+        let indexer = Indexer::from_tokens(&tokens, &locator);
+        let directives = directives::extract_directives(
+            &tokens,
+            directives::Flags::from_settings(&settings),
+            &locator,
+            &indexer,
+        );
+        let LinterResult {
+            data: (diagnostics, _imports),
+            ..
+        } = check_path(
+            Path::new("<filename>"),
+            None,
+            tokens,
+            &locator,
+            &stylist,
+            &indexer,
+            &directives,
+            &settings,
+            flags::Noqa::Enabled,
+        );
+        let actual: Vec<Rule> = diagnostics
+            .into_iter()
+            .map(|diagnostic| diagnostic.kind.rule())
+            .collect();
+        assert_eq!(actual, expected);
+    }
```

Comment:
I copied this from https://github.com/charliermarsh/ruff/blob/main/crates/ruff/src/rules/pandas_vet/mod.rs#L23-L55 - makes me wonder if this was the correct way of doing things, and if so, if there's a nicer way of abstracting these unit tests to keep things in sync. I imagine the general pattern of "this code should raise these linting diagnostics" is quite common.

Reply:
We tend to do this a bit less than might be expected. More commonly, we put the test fixtures in Python files (e.g., `crates/ruff/resources/test/fixtures/flake8_2020`), and then we have a `test_path` utility that runs over files + generates a snapshot. (You could see `rules/flake8_2020/mod.rs` as an example.)

I think this kind of inline testing is fine for simple snippets, if you prefer it. We could DRY up this `rule_code` utility as a TODO for later (and perhaps change it to use snapshots rather than encoding the expected violations).

Reply:
Nice, thanks for that info - planning to open up a follow-on PR documenting how the snapshots work!